include/rocksdb/filter_policy.h

```diff
@@ -80,8 +83,30 @@ class FilterBitsReader {
   }
 };
 
-// Internal type required for FilterPolicy
-struct FilterBuildingContext;
+// Contextual information passed to BloomFilterPolicy at filter building time.
+// Used in overriding FilterPolicy::GetBuilderWithContext().
+struct FilterBuildingContext {
+  // This constructor is for internal use only and subject to change.
+  FilterBuildingContext(const BlockBasedTableOptions& table_options);
+
+  // Returns a new FilterBitsBuilder from the filter_policy in
+  // table_options, or nullptr if not applicable.
+  // (An internal convenience function to save boilerplate.)
+  FilterBitsBuilder* GetBuilder() const;
```

Comment:
It's a little bit odd that a "context" is also a factory class. If it is not too complicated, can we change it to a static function that takes FilterBuildingContext as an argument?

Reply:
Sure. How about a namespace-global function

    FilterBitsBuilder* NewFilterBitsBuilder(const FilterBuildingContext &)

And we could still use temporaries for FilterBuildingContext, as in

    builder = NewFilterBitsBuilder(FilterBuildingContext(table_options));

Except this function will return nullptr if there's no filter policy in effect. Do we have a naming convention for might return a new object or might return nullptr?

Reply:
Actually, I think I'll go with

    builder = BloomFilterPolicy::GetBuilderFromContext(FilterBuildingContext(table_options));

"from context" - The context has everything you need to get the builder.
"with context" - This BloomFilterPolicy creates the builder, consulting information in the context.

And this will be out of view from the public API.